File: 2015/2/main.php
<?php

function parseLine($line)
{
    $line = trim($line);
    $parts = explode('x', $line);
    if (count($parts) == 3) {
        return array_map('intval', $parts);
    }
    throw new Exception('Not enough dimensions.');
}

function readFileAndCalc($filePath, $func)
{
    $fileHandler = fopen($filePath, "r");

    if (!$fileHandler) {
        return "Error opening the file! Check it exists and try again";
    }

    $amountNeeded = 0;
    while (($line = fgets($fileHandler)) !== false) {
        list($length, $width, $height) = parseLine($line);
        // $amountNeeded += calcPresentWrapperPaper($length, $width, $height);
        $amountNeeded += $func($length, $width, $height);
    }
    // Close the file handler
    fclose($fileHandler);

    return $amountNeeded;
}

function calcPresentWrapperPaper($length, $width, $height)
{
    $surfaceArea = 2 * $length * $width + 2 * $width * $height + 2 * $height * $length;
    $smallestSide = min($length * $width, $width * $height, $height * $length);

    return $surfaceArea + $smallestSide;

}

function calcTotalPaperNeeded($filePath)
{
    $totalPaperNeeded = readFileAndCalc($filePath, 'calcPresentWrapperPaper');
    echo "Solution: $totalPaperNeeded square feet of wrapping paper is needed\n";
}

calcTotalPaperNeeded('input.txt');

function calcRibbonNeeded($length, $width, $height)
{
    // Find the two smallest dimensions
    $dimensions = [$length, $width, $height];
    sort($dimensions);
    // Sum of the two smallest sides
    $smallestPerimeter = 2 * ($dimensions[0] + $dimensions[1]);
    $cubic_volume = $length * $width * $height;

    return $smallestPerimeter + $cubic_volume;
}
function calcTotalRibbonNeeded($filePath)
{
    $totalRibbonNeeded = readFileAndCalc($filePath, 'calcRibbonNeeded');
    echo "Solution: $totalRibbonNeeded feet of ribbon is needed\n";
}

calcTotalRibbonNeeded('input.txt');

?>

File: 2015/2/Go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

func convertToIntArray(array []string) ([]int, error) {
	intArray := make([]int, len(array))
	for idx, str := range array {
		num, err := strconv.Atoi(str)
		if err != nil {
			return nil, err
		}
		intArray[idx] = num
	}
	return intArray, nil
}
func parseLine(line string) ([]int, error) {
	line = strings.TrimSpace(line)
	parts := strings.Split(line, "x")
	if len(parts) != 3 {
		return nil, fmt.Errorf("three dimensional string was expected, got %d instead", len(parts))
	}
	return convertToIntArray(parts)

}

func calcPresentWrapperPaper(length int, width int, height int) int {
	bottomTopDim := length * width
	sideDim := width * height
	frontDim := height * length
	surfaceArea := 2*bottomTopDim + 2*sideDim + 2*frontDim
	smallestSide := min(bottomTopDim, sideDim, height*length)
	return surfaceArea + smallestSide
}

func readFileAndCalc(filepath string) (int, int, error) {
	file, err := os.Open(filepath)
	if err != nil {
		fmt.Println("Error opening file", err)
		return 0, 0, err
	}
	defer file.Close()

	totalPaperNeeded := 0  // First challenge
	totalRibbonNeeded := 0 // Second challenge

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		values, err := parseLine(line)
		if err != nil {
			return 0, 0, err
		}
		totalPaperNeeded += calcPresentWrapperPaper(values[0], values[1], values[2])
		totalRibbonNeeded += calcRibbonNeeded(values[0], values[1], values[2])
	}
	return totalPaperNeeded, totalRibbonNeeded, nil
}

func calcRibbonNeeded(length int, width int, height int) int {
	dimensions := []int{length, width, height}
	sort.Ints(dimensions)
	// Sum of the two smallest sides
	smallestPerimeter := 2 * (dimensions[0] + dimensions[1])
	cubicVolume := length * width * height
	return smallestPerimeter + cubicVolume
}

func main() {
	filepath := "input.txt"
	totalPaperNeeded, totalRibbonNeeded, err := readFileAndCalc(filepath)
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	fmt.Printf("Challenge 1: %d square feets of wrapping paper needed\n", totalPaperNeeded)
	fmt.Printf("Challenge 2: %d feets of ribbon needed\n", totalRibbonNeeded)
}
